### DISCUSSION TITLE
Attributes and MASS Services - Custom Lovelace Card

### DISCUSSION THREAD

**Discussion by rxritalin** [music-assistant/support#2533] [Q&A] [answered]
![UI Wip](https://github.com/music-assistant/support/assets/3268460/04bcaf99-c4e8-4170-8382-5796af54198b)
I am in the middle of building out a custom lovelace card utilizing mass. This layout is currently a WIP, so don't be too judgey.
This UI is intended to be for a single player, on a single device and act as the main UI to interface with MASS
At present I am using the "mini-media-player" card, as it seems to offer the most flexibility. 
I can obviously  already use the core features of a media card; 
cover art, 
track
title, 
album
playtime/runtime
Play/Pause, Next, Prev, Shuffle and Repeat

I have already found a simple enough way to access the HASS media player options for Search, PlayLists, Artist, etc and will be building out popup buttons for that shortly. 

This is about where the fun stops though and Im running into issues. 

Issue 1. 
I was hoping to use the Group Management feature built into mini-media-player as it already supports Squeezebox, but when I try to activate a second player, it throws the following error. 
_service.squeezebox.join not found_
I assume that the squeezebox functionality build into MASS does not offer up this service. 
Is there any other way to sync speakers from a lovelace card? Maybe a script? 

Issue 2. 
Current Playlist - I was hoping to be able to list out the currently active queue of music, but I cant seem to find a way to do this. 
I see there is an attribute for active_queue, under the media_player, but I cant seem to find a way to parse this information. 
The output for that attribute is as follows; _active_queue: e0:51:d8:10:95:84_ .
I tried using the mass search function in services, but this does not really work. 
Is there any way to access this data? I would love to be able to build out a scrolling list with selectable items similar to the UI in MASS. 

Issue 3.
Missing Meta Data in Attributes - I don't see any information in the attributes list for current bitrate or container type (OGG, FLAC). 
Not a huge issue but it would be nice to be able to add this info to the UI. 
Play List Name is also missing from the attributes, through I suppose if I can somehow parse the active_queue, it might list there. 

I know there is currently a request for a Lovelace card specifically for MASS, but personally I think a DIY approach would be the better way to go. Just give all the pieces and let people build their own. 
The ytube music player integration in HACs is by far the most feature rich in this regard, but Ill be damned if I'm going back to using YouTube Music. Maybe you can utilize some of his work as a framework for something similar in MASS. The simple access to playlists, current queue, and crazy shit like lyrics are really nice features. 
https://github.com/KoljaWindeler/ytube_music_player

**Comment by OzGav** [comment]
@marcelveldt

**Reply by rxritalin** [reply]
Any thoughts on how to accomplish any of this, or are they things that would need to be added as new features?

**Reply by OzGav** [reply]
Most things would be new features. For issue one, the media player join service call works.

**Reply by OzGav** [reply, marked as answer]
The new `mass.get_library` and `mass.get_queue` actions should provide a solution to the above.

Additionally for issue 2. An API is now available in MA 2.4. You can use curl or `rest_command` from HA to get all the items in the queue. The limit defaults to 500 if you omit it. You are cautioned to not set a value greater then 500 to avoid breaking your system. The practical limit will depend on the resources available on your host. The `player_id` can be found at the top of the player settings page in MA.

```
rest_command:
  ma_queue_test:
    url: http://localhost:8095/api
    method: POST
    headers:
      accept: "application/json, text/html"
    payload: >
      {
        "message_id": "1",
        "command": "player_queues/items",
        "args": {
          "queue_id": "b8:27:eb:8a:b8:8e",
          "limit": 500,
          "offset": 0
        }
      }
    content_type:  'application/json; charset=utf-8'
```

**Comment by OzGav** [comment]
There is now API documentation for the 2.7.0 beta and later versions of Music Assistant. The API documentation is automatically generated and available at http://your_ma_server_ip:8095/api-docs
